test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs

```diff
@@ -782,30 +783,100 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
         }
 
         [Fact]
-        public async Task HeadResponseCanContainContentLengthHeaderButBodyNotWritten()
+        public async Task HeadResponseBodyNotWrittenWithAsyncWrite()
         {
+            var flushed = new SemaphoreSlim(0, 1);
+
             using (var server = new TestServer(async httpContext =>
             {
                 httpContext.Response.ContentLength = 12;
                 await httpContext.Response.WriteAsync("hello, world");
+                await flushed.WaitAsync();
```

Comment:
These calls to WaitAsync() are in the web app on the test. If the web app hangs, the test will not hang. Instead the test will fail when it times out waiting for a response. The response timeout here is 1 minute like for most of our tests.